### PR TITLE
Spaces.includeSubSpaceRoomsInRoomList settings

### PR DESCRIPTION
**Pull request by kerryarchibald**
Adds a setting to allow only showing direct children of a space in the room list:
Screenshot 2022-01-31 at 17 37 58
Screenshot 2022-01-31 at 17 38 39
Screenshot 2022-01-31 at 17 38 54

---
Here's what your changelog entry will look like:

## ✨ Features
 * Spaces.includeSubSpaceRoomsInRoomList settings ([\#7682](https://github.com/matrix-org/matrix-react-sdk/pull/7682)). Contributed by @kerryarchibald.

Preview: https://pr7682--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/settings/Settings.tsx [hunk 1]
```diff
@@ -858,6 +858,11 @@ export const SETTINGS: {[setting: string]: ISetting} = {
         default: true,
         controller: new IncompatibleController("showCommunitiesInsteadOfSpaces", null),
     },
+    "Spaces.includeSubSpaceRoomsInRoomList": {
+        displayName: _td("Include all sub-space rooms in Space room list"),
```
Comment (t3chguy) [inline src/settings/Settings.tsx:862]: We consider sub-space/subspace jargon and do not use it in the UI

File: test/stores/room-list/filters/SpaceFilterCondition-test.ts [hunk 2]
```diff
@@ -0,0 +1,237 @@
+import { mocked } from 'ts-jest/utils';
```
Comment (t3chguy) [inline test/stores/room-list/filters/SpaceFilterCondition-test.ts:1]: copyright

File: src/stores/spaces/SpaceStore.ts [hunk 3]
```diff
@@ -101,6 +101,8 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
     private notificationStateMap = new Map<SpaceKey, SpaceNotificationState>();
     // Map from space key to Set of room IDs that should be shown as part of that space's filter
     private spaceFilteredRooms = new Map<SpaceKey, Set<string>>(); // won't contain MetaSpace.People
+    // Map from space key to Set of room IDs that should be shown as part of that space's filter
+    private spaceFilteredDirectChildRooms = new Map<SpaceKey, Set<string>>(); // won't contain MetaSpace.People
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:105]: also presumably does not need to contain the other metaspaces given that'd just be duplicated identical data

File: src/stores/spaces/SpaceStore.ts [hunk 4]
```diff
@@ -387,6 +393,13 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
         return this.spaceFilteredRooms.get(space) || new Set();
     };
 
+    public getSpaceFilteredDirectChildRoomIds = (space: SpaceKey): Set<string> => {
+        if (space === MetaSpace.Home && this.allRoomsInHome) {
+            return this.getSpaceFilteredRoomIds(space);
+        }
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:399]: IMO all metaspaces should get handled by the non-direct variant given metaspaces can't have child spaces

File: src/stores/spaces/SpaceStore.ts [hunk 5]
```diff
@@ -719,7 +733,13 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
                 const expandedRoomIds = new Set(Array.from(roomIds).flatMap(roomId => {
                     return this.matrixClient.getRoomUpgradeHistory(roomId, true).map(r => r.roomId);
                 }));
+                const expandedDirectChildRoomIds = new Set(directChildRoomIds.flatMap(roomId => {
+                    return this.matrixClient.getRoomUpgradeHistory(roomId, true).map(r => r.roomId);
+                }));
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:738]: this duplicates work between the two loops

**Review by t3chguy — COMMENTED** [review]
This implementation, whilst functional, pushes into the hard-to-maintain. I think this might be a point where refactoring the SpaceFilterCondition and SpaceNotificationState to be hierarchical (contain child instances of themselves for each sub-space) and then only store Maps from spaceId->direct children would be more sane as then simply culling the hierarchical behaviour in SpaceFilterCondition would achieve this setting

File: src/stores/spaces/SpaceStore.ts (hunk 3 above)
Comment (t3chguy) [reply src/stores/spaces/SpaceStore.ts:105]: Not only does not need to contain, simply would not contain, given you only populate it with spaceIds, so the signature needs updating to reflect reality

File: src/settings/Settings.tsx (hunk 1 above)
Comment (kerryarchibald) [reply src/settings/Settings.tsx:862]: What language should be used? is there a guide for this somewhere?

**Comment by kerryarchibald** [comment]
@t3chguy Is it the amount of state tracked to check for changes in `onStoreUpdate` that concerns you?
Adding hierarchical space filters seems like adding quite a lot of complexity for a relatively small win. Maintenance of both direct and indirect child lists in `SpaceStore` is still needed as `isRoomInSpace` depends on the indirect child list and is used pretty widely. It also complexifies the application of space settings - should the active spaces settings be applied down the hierarchy?

File: src/settings/Settings.tsx (hunk 1 above)
Comment (t3chguy) [reply src/settings/Settings.tsx:862]: There is no guide given that Subspaces are still in Beta.

image

There is no existing language here but all instances of "sub-space" were previously stripped from the app so should not be re-introduced without being done holistically and intentionally.

**Comment by t3chguy** [comment]
> @t3chguy Is it the amount of state tracked to check for changes in `onStoreUpdate` that concerns you? Adding hierarchical space filters seems like adding quite a lot of complexity for a relatively small win. Maintenance of both direct and indirect child lists in `SpaceStore` is still needed as `isRoomInSpace` depends on the indirect child list and is used pretty widely. It also complexifies the application of space settings - should the active spaces settings be applied down the hierarchy?

The thing that concerns me is the very minute differences between the various fields, which would go away if things were done hierarchically, like how this new field is indirect children except doesn't concern itself with metaspaces, etc.

**Comment by t3chguy** [comment]
How does this interact with notification badges? If you set it to false then will the notifications badge still aggregate counts of grand*-child rooms?

**Comment by t3chguy** [comment]
> Hierarchical space filters hit a major snag on cycles being allowed in space hierarchy. Shelving that for now to get this feature over the line.

Sure, the typical approach we take is keeping a parentPath Set to prevent hitting upon such a cycle, would be nice if you could stick up what you have into a draft PR so someone else might continue it if you can't :)

**Comment by kerryarchibald** [comment]
Notification states when sub space rooms are hidden are a bit strange. I don't use any other messaging tools with the level of nesting that Element allows so I'm not sure what behaviour is usual. @nadonomy what do you think?

https://user-images.githubusercontent.com/3055605/154673154-d29de9de-90c3-4c5c-adec-b0622e177e4b.mov

Expanding the space panel is ok, but makes it at least 3 clicks to find your unread room. 
Screenshot 2022-02-18 at 12 08 24
Screenshot 2022-02-18 at 12 08 32

Compared to probably 2 clicks in the current room list:
Screenshot 2022-02-18 at 12 24 54

File: src/stores/spaces/SpaceStore.ts [hunk 6]
```diff
@@ -136,6 +136,7 @@ export class SpaceStoreClass extends AsyncStoreWithClient<IState> {
         SettingsStore.monitorSetting("Spaces.allRoomsInHome", null);
         SettingsStore.monitorSetting("Spaces.enabledMetaSpaces", null);
         SettingsStore.monitorSetting("Spaces.showPeopleInSpace", null);
+        SettingsStore.monitorSetting("Spaces.includeSubSpaceRoomsInRoomList", null);
```
Comment (t3chguy) [inline src/stores/spaces/SpaceStore.ts:139]: Why is this monitoring a setting which it doesn't react to changes of?

**Review by t3chguy — CHANGES_REQUESTED** [review]
`switchToRelatedSpace` will also need updating to handle the case where the root space has `Spaces.includeSubSpaceRoomsInRoomList` set to `false` and yet it still will choose it but the room will not be visible in the filtered Room List.

**Comment by QuditWolf** [comment]
Seems like a really niche feature add!
What are the odds that this still gets merged?


My active subspace rooms flood the space rooms and I can't find them without searching for them...

**Pull request closed without merging**